tools/autograd/gen_variable_type.py

```diff
@@ -409,14 +410,29 @@ def gen_variable_type_shard(
         with native_function_manager(f):
             name = cpp.name(f.func)
             formals = gen_formals(f)
-
-            type_definitions.append(METHOD_DEFINITION.substitute(
-                return_type=cpp.returns_type(f.func.returns).cpp_type(),
-                type_wrapper_name=type_wrapper_name(f),
-                type_definition_body=emit_body(fn),
-                formals=formals,
-            ))
-            wrapper_registrations.append(gen_wrapper_registration(f))
+            # NOTE: [ Registering AutogradNotImplemented boxed kernel ]
+            #
+            # When there is no derivatives.yaml entry, we register a generic boxed NotImplemented kernel
+            # to set grad_fn to be NotImplemented, so that forward proceeds as usual but an error is
+            # properly produced on backward.
+            #
+            # There are two cases where still let codegen handle it:
+            # 1) ops that need to reset grad accumulator (we let codegen handle this case because) the
+            #    list is (currently) only accessible in Python and there are only 2 op.
+            # 2) User explicitly specifies DONT_REQUIRE_DERIVATIVE. This basically makes autograd a
+            #    fallthrough with NDEBUG checks. This can be useful for when all outputs are integral.
+            #
+            if fn.info is None and not get_base_name(f) in RESET_GRAD_ACCUMULATOR and not get_base_name(f) in DONT_REQUIRE_DERIVATIVE:
+                wrapper_registrations.append(AUTOGRAD_NOT_IMPLEMENTED_REGISTRATION.substitute(
+                    unqual_operator_name_with_overload=f.func.name))
+            else:
+                type_definitions.append(METHOD_DEFINITION.substitute(
+                    return_type=cpp.returns_type(f.func.returns).cpp_type(),
+                    type_wrapper_name=type_wrapper_name(f),
+                    type_definition_body=emit_body(fn),
+                    formals=formals,
+                ))
+                wrapper_registrations.append(gen_wrapper_registration(f))
```

Comment:
I highly recommend doing this in a separate PR (although I suppose this is good for testing right now). In particular, I'm doubtful you actually want to default not implemented to the boxed fallback, esp if you have a templated one coming down the pipes.